### PR TITLE
Loop optimization

### PR DESCRIPTION
**Pull request by evkotov**
### Details:
 - Optimize loop after TF FE

### Tickets:
 - 136813

### CODE REVIEW

File: src/common/transformations/include/transformations/common_optimizations/loop_inputs.hpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+namespace pass {
+
+class TRANSFORMATIONS_API LoopInputs;
+
+}  // namespace pass
+}  // namespace ov
+
+/**
+ * @ingroup ov_transformation_common_api
+ * @brief LoopInputs transformation manage Loop inputs/outputs
```
Comment (itikhono) [inline src/common/transformations/include/transformations/common_optimizations/loop_inputs.hpp:23]: please provide more details in the description

File: src/common/transformations/include/transformations/common_optimizations/loop_inputs.hpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+class ov::pass::LoopInputs : public ov::pass::MatcherPass {
+public:
+    OPENVINO_RTTI("ConcatFusion", "0");
```
Comment (itikhono) [inline src/common/transformations/include/transformations/common_optimizations/loop_inputs.hpp:28]: ConcatFusion?

File: src/common/transformations/include/transformations/common_optimizations/loop_inputs.hpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+#include <memory>
+#include <vector>
+
+#include "openvino/pass/graph_rewrite.hpp"
+#include "transformations_visibility.hpp"
+
+namespace ov {
+namespace pass {
+
+class TRANSFORMATIONS_API LoopInputs;
```
Comment (itikhono) [inline src/common/transformations/include/transformations/common_optimizations/loop_inputs.hpp:16]: LoopInputs does not reflect information on modifications performed in the transformation
Please rename it

File: src/common/transformations/src/transformations/common_optimizations/loop_inputs.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+namespace {
+std::shared_ptr<ov::op::v0::Parameter> get_parent_param(const std::shared_ptr<ov::op::v0::Result>& result) {
+    return ov::as_type_ptr<ov::op::v0::Parameter>(result->input_values()[0].get_node_shared_ptr());
+}
+}  // namespace
+
+ov::pass::LoopInputs::LoopInputs() {
+    MATCHER_SCOPE(LoopInputs);
+
+    auto loop_label = wrap_type<ov::op::v5::Loop>();
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/loop_inputs.cpp:30]: can we extend it for TensorIterator?

File: src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp [hunk 5]
```diff
@@ -148,6 +149,7 @@ bool ov::pass::MOCTransformations::run_on_model(const std::shared_ptr<ov::Model>
     // the order EliminateDuplicateTIInputs and RemoveMultiSubGraphOpDanglingParamsResults is important
     // it looks like we need to combine these transformations into one.
     REGISTER_PASS(manager, EliminateDuplicateTIInputs);
+    REGISTER_PASS(manager, LoopInputs);
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp:152]: maybe it's better to move it before EliminateDuplicateTIInputs

File: src/common/transformations/src/transformations/common_optimizations/loop_inputs.cpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+        std::unordered_map<size_t, ov::Output<ov::Node>> loop_inputs;  // body_param->instance_id -> loop_input
+        for (const auto& input_description : loop->get_input_descriptions()) {
+            loop_inputs.emplace(body_params[input_description->m_body_parameter_index]->get_instance_id(),
+                                loop_input_values[input_description->m_input_index]);
+        }
+
+        ov::OutputVector new_loop_outputs;
+        for (const auto& output_description : loop->get_output_descriptions()) {
+            int64_t iteration = -1;
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/loop_inputs.cpp:91]: we don't need to create it multiple times
Please move it above or use the value in the function call

File: src/common/transformations/src/transformations/common_optimizations/loop_inputs.cpp [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        const auto& trip_count = loop_input_values[0];
+        const auto& exec_cond = loop_input_values[1];
+
+        auto new_loop = make_shared<ov::op::v5::Loop>(trip_count, exec_cond);
+        new_loop->set_function(body_model);
+        new_loop->set_special_body_ports(loop->get_special_body_ports());
+
+        for (const auto& input_description : loop->get_input_descriptions()) {
+            if (const auto merged_input_desc =
+                    dynamic_pointer_cast<ov::op::util::MultiSubGraphOp::MergedInputDescription>(input_description)) {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/loop_inputs.cpp:50]: mixing as_type_ptr (line 35)  and dynamic_pointer_cast (line 50)
please select one method

File: src/common/transformations/src/transformations/common_optimizations/loop_inputs.cpp [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                if (get_parent_param(body_results[merged_input_desc->m_body_value_index])) {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/loop_inputs.cpp:51]: please add a comment explaining the logic here
I believe we have to check that body_params[merged_input_desc->m_body_parameter_index] is exactly the parameter that `get_parent_param`  method returns.

File: src/common/transformations/include/transformations/common_optimizations/loop_inputs.hpp (hunk 1 above)
Comment (rkazants) [inline src/common/transformations/include/transformations/common_optimizations/loop_inputs.hpp:23]:
```suggestion
 * @brief LoopInputs transformation manages Loop inputs/outputs
```

File: src/common/transformations/src/transformations/common_optimizations/loop_inputs.cpp [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        ov::OutputVector new_loop_outputs;
+        for (const auto& output_description : loop->get_output_descriptions()) {
+            int64_t iteration = -1;
+            if (const auto body_output_desc =
+                    dynamic_pointer_cast<ov::op::util::MultiSubGraphOp::BodyOutputDescription>(output_description)) {
+                iteration = body_output_desc->m_iteration;
+            }
+            const auto& body_result = body_results[output_description->m_body_value_index];
+            if (const auto& body_param = get_parent_param(body_result)) {
+                new_loop_outputs.emplace_back(loop_inputs[body_param->get_instance_id()]);
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/loop_inputs.cpp:98]: It looks like we need to clean up Input Descriptions
Maybe it's better to combine processing Input and Output Descriptions

File: src/common/transformations/include/transformations/common_optimizations/loop_inputs.hpp (hunk 3 above)
Comment (rkazants) [inline src/common/transformations/include/transformations/common_optimizations/loop_inputs.hpp:16]:
```suggestion
class TRANSFORMATIONS_API EliminateLoopInputsOutputs;
```

**Review by itikhono — CHANGES_REQUESTED** [review]

**Comment by itikhono** [comment]
please provide more details in the description section 
Details:
Optimize loop after TF FE

File: src/common/transformations/src/transformations/common_optimizations/loop_inputs.cpp (hunk 8 above)
Comment (evkotov) [reply src/common/transformations/src/transformations/common_optimizations/loop_inputs.cpp:51]: function get_parent_param() returns Parameter, that is actually parent node to argument. If parent node is not Parameter, this function returns empty shared_ptr

File: src/common/transformations/src/transformations/common_optimizations/eliminate_loop_inputs_outputs.cpp [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+    auto subgraph_label = wrap_type<ov::op::v5::Loop, ov::op::v0::TensorIterator>();
+
+    matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        const auto& pattern_to_output = m.get_pattern_value_map();
+
+        auto subgraph = as_type_ptr<op::util::SubGraphOp>(pattern_to_output.at(subgraph_label).get_node_shared_ptr());
+        auto loop = as_type_ptr<ov::op::v5::Loop>(subgraph);
+
+        // clone model since we will update it for the new loop
+        auto body_model = subgraph->get_function()->clone();
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/eliminate_loop_inputs_outputs.cpp:187]: please do not clone the original model

**Pull request closed without merging**